[libc++][functional] Implement `not_fn<NTTP>`

JMazurkiewicz opened the pull request:
Implement `not_fn<NTTP>` function from ["P2714R1 Bind front and back to NTTP callables"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2714r1.html).

JMazurkiewicz committed:
[libc++] Implement `not_fn<NTTP>`

Implement `not_fn<NTTP>` function from "P2714R1 Bind front and back to NTTP callables".

xiaoyang-sde commented on libcxx/include/__functional/not_fn.h:
```diff
@@ -48,6 +50,27 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 auto not_fn(_Fn&& __f) {
+template <auto _Fn>
+struct __nttp_not_fn_t {
+  template <class... _Args>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Args&&... __args) const
```
```suggestion
  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto operator()(_Args&&... __args)
```

I think we can make this operator `static`?

frederick-vs-ja replied:
`volatile` values can be rejected by not using `static`, but this doesn't seem worthwhile... See [[func.require]/4](https://eel.is/c++draft/func.require#4) and [LWG4007](https://cplusplus.github.io/LWG/issue4007).

JMazurkiewicz replied:
Initially, I've made `operator()` static but due to the reason previously mentioned

> volatile values can be rejected by not using static

I've made it member function again. For now, I'm going to leave it as-is and wait for maintainers' comment.

ldionne commented on libcxx/test/libcxx/utilities/function.objects/func.not.fn/not_fn.nttp.compile.pass.cpp:
```diff
@@ -0,0 +1,29 @@
+};
+
+void test() {
+  using ResultWithEmptyFuncObject = decltype(std::not_fn<std::false_type{}>());
+  static_assert(std::is_empty_v<ResultWithEmptyFuncObject>);
+
+  using ResultWithNotEmptyFuncObject = decltype(std::not_fn<NonEmptyFunctionObject{}>());
+  static_assert(std::is_empty_v<ResultWithNotEmptyFuncObject>);
```
This doesn't need to be in a function, it can be at file scope!

ldionne commented on libcxx/test/libcxx/utilities/function.objects/func.not.fn/not_fn.nttp.compile.pass.cpp:
```diff
@@ -0,0 +1,29 @@
+#include <functional>
+#include <type_traits>
+
+struct NonEmptyFunctionObject {
```
I'd also test with a couple of other callables like a function pointer and a pointer to member function.

ldionne commented on libcxx/test/libcxx/utilities/function.objects/func.not.fn/not_fn.nttp.nodiscard.verify.cpp:
```diff
@@ -0,0 +1,21 @@
+
+void test() {
+  using F = std::true_type;
+  std::not_fn<F{}>(); // expected-warning {{ignoring return value of function declared with 'nodiscard' attribute}}
```
We should also be testing that actually calling the `not_fn` is nodiscard, since you marked it as such.

```
auto negated = std::not_fn<F{}>();
negated(); // expected-warning {{ignoring return value of function declared with 'nodiscard' attribute}}
```

ldionne commented on libcxx/test/std/utilities/function.objects/func.not_fn/not_fn.nttp.pass.cpp:
```diff
@@ -0,0 +1,294 @@
+    assert(static_cast<const F&&>(f)() == -2);
+  }
+
+  // Call to `not_fn<NTTP>` unspecified-type's operator() should always result in call to const& overload of .
```
Unfinished sentence?

ldionne commented on libcxx/test/std/utilities/function.objects/func.not_fn/not_fn.nttp.pass.cpp:
```diff
@@ -0,0 +1,294 @@
+
+  // Call to `not_fn<NTTP>` unspecified-type's operator() should always result in call to const& overload of .
+  {
+    { // Make sure unspecified-type is still callable when we delete & overload.
```
```suggestion
    { // Make sure unspecified-type is still callable when we delete the & overload.
```

Throughout below.

ldionne reviewed (changes requested):
Thanks for the patch and sorry for this falling under the radar. I have a few comments but this looks mostly good. The tests are really comprehensive, thanks for that!

JMazurkiewicz committed:
Move assertions to file scope

Comment: https://github.com/llvm/llvm-project/pull/86133#discussion_r1862658517

JMazurkiewicz committed:
Test more `not_fn`s for emptiness

Comment: https://github.com/llvm/llvm-project/pull/86133#discussion_r1862659285

JMazurkiewicz committed:
Add missing `[[nodiscard]]` test

Comment: https://github.com/llvm/llvm-project/pull/86133#discussion_r1862660535

JMazurkiewicz committed:
the

Comment: https://github.com/llvm/llvm-project/pull/86133#discussion_r1862662233

JMazurkiewicz committed:
Add extra `noexcept` test

JMazurkiewicz committed:
Finish sentence

Comment: https://github.com/llvm/llvm-project/pull/86133#discussion_r1862662065

ldionne reviewed (approved):
LGTM, thanks!

ldionne commented:
You probably want to rebase onto `main` and re-push to get the CI going again. Once it's green this is good to merge.

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/nttp_not_fn

JMazurkiewicz committed:
Fix ASAN/TSAN/UBSAN/MSAN CI

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/nttp_not_fn

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/nttp_not_fn

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/nttp_not_fn
